File: cs/ch17-04-arithmetic/Money/Operation.cs


namespace TheSoftwareGorilla.TDD.Money;

public class Operation<T> : IOperation<T> where T : ICurrencyHolder<T>
{   
    public T LeftOperand { get; }
    public T RightOperand { get; }

    public Bank<T> Bank { get; }

    private readonly Func<T, T, T> _calculation;
    private readonly string _toCurrency;


    public Operation(T leftOperand, T rightOperand, Bank<T> bank, string toCurrency, Func<T, T, T> calculation)
    {
        LeftOperand = leftOperand;
        RightOperand = rightOperand;
        Bank = bank;
        _calculation = calculation;
        _toCurrency = toCurrency;
    }

    public virtual T Evaluate()
    {
        T left = LeftOperand.Convert(_toCurrency);
        T right = RightOperand.Convert(_toCurrency); 
        return _calculation(left, right);
    }

}

File: java/ch17-04-arithmetic/src/main/java/com/thesoftwaregorilla/tdd/money/Operation.java
package com.thesoftwaregorilla.tdd.money;

import java.util.function.BiFunction;

public class Operation<T extends ICurrencyHolder<T>> implements IOperation<T> {

    private final T leftOperand;
    private final T rightOperand;
    private final Bank<T> bank;
    private final String toCurrency;
    private final BiFunction<T, T, T> calculation;

    public Operation(T leftOperand, T rightOperand, Bank<T> bank, String toCurrency, BiFunction<T, T, T> calculation) {
        this.leftOperand = leftOperand;
        this.rightOperand = rightOperand;
        this.bank = bank;
        this.calculation = calculation;
        this.toCurrency = toCurrency;
    }

    public T evaluate() {
        T left = leftOperand.convert(toCurrency);
        T right = rightOperand.convert(toCurrency);
        return calculation.apply(left, right);
    }

    public T getLeftOperand() {
        return leftOperand;
    }

    public T getRightOperand() {
        return rightOperand;
    }

    public Bank<T> getBank() {
        return bank;
    }

}
